node/src/Pos/Client/Txp/Util.hs

```diff
@@ -261,85 +323,97 @@ groupUtxo utxo =
     -> TxFee
     -> TxCreator m TxRaw
-prepareTxRaw utxo outputs (TxFee fee) = do
+prepareTxRawWithPicker inputPicker utxo outputs (TxFee fee) = do
     mapM_ (checkIsNotRedeemAddr . txOutAddress . toaOut) outputs
 
     totalMoney <- sumTxOuts outputs
     when (totalMoney == mkCoin 0) $
         throwError $ GeneralTxError "Attempted to send 0 money"
 
-    let totalMoneyWithFee = totalMoney `unsafeAddCoin` fee
-    futxo <- either throwError pure $
-        evalStateT (pickInputs []) (InputPickerState totalMoneyWithFee sortedGroups)
+    let moneyToSpent = totalMoney `unsafeAddCoin` fee
+    futxo <- either throwError pure $ inputPicker utxo outputs moneyToSpent
     case nonEmpty futxo of
         Nothing       -> throwError $ GeneralTxError "Failed to prepare inputs!"
         Just inputsNE -> do
             totalTxAmount <- sumTxOuts $ map snd inputsNE
             let trInputs = map formTxInputs inputsNE
-                trRemainingMoney = totalTxAmount `unsafeSubCoin` totalMoneyWithFee
+                trRemainingMoney = totalTxAmount `unsafeSubCoin` moneyToSpent
             let trOutputs = outputs
             pure TxRaw {..}
   where
     sumTxOuts = either (throwError . GeneralTxError) pure .
         integerToCoin . sumTxOutCoins
-    gUtxo = groupUtxo utxo
-    outputAddrsSet = foldl' (flip HS.insert) mempty $
-        map (txOutAddress . toaOut) outputs
-    isOutputAddr = flip HS.member outputAddrsSet
-    sortedGroups = sortOn (Down . ugTotalMoney) $
-        filter (not . isOutputAddr . ugAddr) gUtxo
-    disallowedInputGroups = filter (isOutputAddr . ugAddr) gUtxo
-    disallowedMoney = sumCoins $ map ugTotalMoney disallowedInputGroups
-
-    pickInputs :: FlatUtxo -> InputPicker FlatUtxo
-    pickInputs inps = do
-        moneyLeft <- use ipsMoneyLeft
-        if moneyLeft == mkCoin 0
-            then return inps
-            else do
-                mNextOutGroup <- head <$> use ipsAvailableOutputGroups
-                case mNextOutGroup of
-                    Nothing -> if disallowedMoney >= coinToInteger moneyLeft
-                        then throwError $ NotEnoughAllowedMoney moneyLeft
-                        else throwError $ NotEnoughMoney moneyLeft
-                    Just UtxoGroup {..} -> do
-                        ipsMoneyLeft .= unsafeSubCoin moneyLeft (min ugTotalMoney moneyLeft)
-                        ipsAvailableOutputGroups %= tail
-                        pickInputs (toList ugUtxo ++ inps)
-
     formTxInputs (inp, TxOutAux txOut) = (txOut, inp)
-
     checkIsNotRedeemAddr outAddr =
         when (isRedeemAddress outAddr) $
             throwError $ OutputIsRedeem outAddr
 
+prepareTxRaw
+    :: Monad m
+    => Utxo
+    -> TxOutputs
+    -> TxFee
+    -> TxCreator m TxRaw
+prepareTxRaw utxo outputs fee = do
+    UseGroupedInputs whetherGrouped <- view tcdUseGroupedInputs
+    let inputPicker =
+            if whetherGrouped
+            then groupedInputPicker
+            else plainInputPicker
```

Comment:
So this `if` would be replaced by a `case` on the input selection policy. This way it is naturally extendible when we add more different policy choices.